modules/compliance-tests/src/smithy4s/compliancetests/weaverimpl/ProtocolComplianceTestSuite.scala

```diff
@@ -0,0 +1,200 @@
+/*
+ *  Copyright 2021-2022 Disney Streaming
+ *
+ *  Licensed under the Tomorrow Open Source Technology License, Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     https://disneystreaming.github.io/TOST-1.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s.compliancetests.weaverimpl
+
+import smithy4s.compliancetests.{ComplianceTest, ReverseRouter}
+import smithy4s.compliancetests.internals.AllowRules
+import smithy4s.{Document, Schema, ShapeId}
+import smithy4s.dynamic.DynamicSchemaIndex
+import cats.syntax.all._
+import cats.effect.IO
+import cats.effect.std.Env
+import fs2.io.file.Path
+import smithy4s.http.CodecAPI
+import smithy4s.dynamic.model.Model
+import smithy4s.dynamic.DynamicSchemaIndex.load
+import smithy4s.schema.Schema.document
+import smithy4s.compliancetests._
+import smithy4s.http.PayloadError
+import weaver._
+
+abstract class ProtocolComplianceTestSuite
+    extends EffectSuite[IO]
+    with BaseCatsSuite {
+
+  implicit protected def effectCompat: EffectCompat[IO] = CatsUnsafeRun
+
+  def getSuite: EffectSuite[IO] = this
+
+  def allowAll: Boolean = true
+
+  def allTests: DynamicSchemaIndex => List[ComplianceTest[IO]]
+
+  def spec(args: List[String]): fs2.Stream[IO, TestOutcome] = {
+    fs2.Stream
+      .eval(dynamicSchemaIndexLoader)
+      .evalMap(index => decodeAllowRules(index).map(index -> _))
+      .flatMap { case (schemaIndex, allowRules) =>
+        fs2.Stream.emits(allTests(schemaIndex).map((allowRules, _)))
+      }
+      .evalMap { case (allowRules, test) =>
+        runInWeaver(allowAll, allowRules, test)
+      }
+  }
+
+  def dynamicSchemaIndexLoader: IO[DynamicSchemaIndex]
+
+  def genClientTests(
+      impl: ReverseRouter[IO],
+      shapeIds: ShapeId*
+  ): DynamicSchemaIndex => List[ComplianceTest[IO]] = { dynamicSchemaIndex =>
```

Comment:
Make this a curried method as opposed to returning a function. 

The rule of thumb is that if the first thing you write in the body of your method is the lambda, then you should make a curried method instead : 

```
def genClientTests(impl: ReverseRouter[IO], shapeIds: ShapeId*)(dsi: DynamicSchemaIndex) : List[ComplianceTest[IO]] = {...}  
```